capstone/src/main/webapp/visualizationdata.test.js

```diff
@@ -0,0 +1,235 @@
+  // Empty dataset provided
+  expect(functions.getFirstDate([]))
+      .toMatchObject({'minWeek': null, 'minYear': null});
+
+  // Tests with different month/year edge cases
+  expect(functions.getFirstDate(pageViewExample1))
+      .toMatchObject({'minWeek': 10, 'minYear': 2020});
+  expect(functions.getFirstDate(pageViewExample2))
+      .toMatchObject({'minWeek': 11, 'minYear': 2020});
+  expect(functions.getFirstDate(pageViewExample3))
+      .toMatchObject({'minWeek': 25, 'minYear': 2019});
+});
+
+/**
+ * Test for getDateFromWeekYear to ensure it finds the correct Date from
+ * January 1 based on checking the calendar. All dates are Sundays of the
+ * week requested.
+ */
+test('Test getDateFromWeekYear with different inputs', () => {
+  expect(functions.getDateFromWeekYear(4, 2020))
+      .toMatchObject(new Date(2020, 0, 19));
+  expect(functions.getDateFromWeekYear(1, 2020))
+      .toMatchObject(new Date(2019, 11, 29));
+  expect(functions.getDateFromWeekYear(26, 2020))
+      .toMatchObject(new Date(2020, 5, 21));
+
+  // Return null for negative year or week values
+  expect(functions.getDateFromWeekYear(-1, 2020)).toBe(null);
+  expect(functions.getDateFromWeekYear(2, -50)).toBe(null);
+});
+
+/**
+ * Test for getFullDateArray with different cases of crossing across
+ * years/months, etc
+ */
+test('Test getFullDateArray with different inputs', () => {
+  const expected1 = [new Date(2020, 6, 15), new Date(2020, 6, 22)];
+  expect(
+      functions.getFullDateArray(new Date(2020, 6, 15), new Date(2020, 6, 22)))
+      .toMatchObject(expected1);
+  const expected2 =
+      [new Date(2020, 6, 1), new Date(2020, 6, 8), new Date(2020, 6, 15)];
+  expect(
+      functions.getFullDateArray(new Date(2020, 6, 1), new Date(2020, 6, 21)))
+      .toMatchObject(expected2);
+  const expected3 =
+      [new Date(2019, 11, 25), new Date(2020, 0, 1), new Date(2020, 0, 8)];
+  expect(
+      functions.getFullDateArray(new Date(2019, 11, 25), new Date(2020, 0, 9)))
+      .toMatchObject(expected3);
+});
+
+/**
+ * Test getNumWeeksBetween with different edge cases of same dates, crossing
+ * years/months, etc
+ */
+test('Test getNumWeeksBetween with different inputs', () => {
+  expect(functions.getNumWeeksBetween(
+             new Date(2020, 6, 21), new Date(2020, 6, 28)))
+      .toBe(1);
+  expect(
+      functions.getNumWeeksBetween(new Date(2020, 6, 21), new Date(2020, 7, 4)))
+      .toBe(2);
+  expect(functions.getNumWeeksBetween(
+             new Date(2020, 6, 21), new Date(2020, 6, 21)))
+      .toBe(0);
+  expect(functions.getNumWeeksBetween(
+             new Date(2019, 11, 22), new Date(2020, 0, 19)))
+      .toBe(4);
+
+  // Should work in reverse param order also
+  expect(functions.getNumWeeksBetween(
+             new Date(2020, 6, 28), new Date(2020, 6, 21)))
+      .toBe(1);
+  expect(
+      functions.getNumWeeksBetween(new Date(2020, 7, 4), new Date(2020, 6, 21)))
+      .toBe(2);
+});
```

Comment:
Consider creating a test for each input. That way when one of the tests fails, it's super easy to figure out which case failed (since the test name would describe the exact input being tested). (Here and for the other tests with multiple cases.)

Reply:
Sounds great, I separated all the tests!